File: store/carousel_admin.py
from django.contrib import admin
from django import forms
from django.utils.html import format_html
from .models import CarouselSlide

@admin.register(CarouselSlide)
class CarouselSlideAdmin(admin.ModelAdmin):
    list_display = ['title', 'slide_type', 'image_preview', 'is_active', 'order', 'created_at']
    list_filter = ['slide_type', 'is_active', 'created_at']
    search_fields = ['title', 'subtitle', 'description']
    list_editable = ['is_active', 'order']
    ordering = ['order', '-created_at']
    
    fieldsets = (
        ('Información Principal', {
            'fields': ('title', 'subtitle', 'description', 'slide_type', 'is_active', 'order')
        }),
        ('Imagen del Slide', {
            'fields': ('image', 'image_preview_large'),
            'description': 'Sube una imagen para el slide del carrusel'
        }),
        ('Botón de Acción', {
            'fields': ('button_text', 'button_link'),
            'classes': ('collapse',),
        }),
        ('Personalización', {
            'fields': ('background_color', 'text_color'),
            'classes': ('collapse',),
        }),
    )
    
    readonly_fields = ['image_preview_large']
    
    def image_preview(self, obj):
        """Miniatura de la imagen en la lista"""
        if obj.image:
            return format_html(
                '<img src="{}" style="width: 80px; height: 40px; object-fit: cover; border-radius: 4px; box-shadow: 0 2px 4px rgba(0,0,0,0.1);" />',
                obj.image.url
            )
        return format_html('<span style="color: #999;">Sin imagen</span>')
    image_preview.short_description = 'Vista Previa'
    
    def image_preview_large(self, obj):
        """Vista previa grande en el formulario de edición"""
        if obj.image:
            return format_html(
                '''
                <div style="margin-top: 10px; padding: 15px; background: #f8f9fa; border-radius: 8px; border: 1px solid #dee2e6;">
                    <p style="margin: 0 0 10px 0; font-weight: bold; color: #495057;">📷 Imagen actual del slide:</p>
                    <img src="{}" style="max-width: 500px; max-height: 300px; object-fit: contain; border: 2px solid #dee2e6; border-radius: 8px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);" />
                    <p style="margin: 10px 0 0 0; font-size: 12px; color: #6c757d;">
                        <strong>URL:</strong> <a href="{}" target="_blank">{}</a>
                    </p>
                </div>
                ''',
                obj.image.url,
                obj.image.url,
                obj.image.url
            )
        return format_html('<p style="color: #999; font-style: italic;">No hay imagen cargada aún</p>')
    image_preview_large.short_description = 'Vista Previa Actual'
    
    def save_model(self, request, obj, form, change):
        """Override para debugging al guardar"""
        print(f"\n{'='*50}")
        print(f"🔄 Guardando CarouselSlide: {obj.title}")
        print(f"{'='*50}")
        
        # Verificar si hay imagen
        if 'image' in form.cleaned_data and form.cleaned_data['image']:
            print(f"✅ Imagen detectada: {form.cleaned_data['image']}")
        else:
            print(f"⚠️ No se detectó imagen nueva")
        
        # Guardar el objeto
        super().save_model(request, obj, form, change)
        
        # Verificar que se guardó
        if obj.image:
            print(f"✅ Imagen guardada exitosamente: {obj.image.url}")
        else:
            print(f"❌ No se guardó ninguna imagen")
        print(f"{'='*50}\n")
    
    class Media:
        css = {
            'all': ('admin/css/carousel_admin.css',)
        }
        js = ('admin/js/carousel_admin.js',)

print("✅ CarouselSlideAdmin registrado correctamente")

File: static/admin/js/carousel_admin.js
// JavaScript para mejorar el admin del carrusel
(function($) {
    'use strict';
    
    // Verificar que jQuery está disponible
    if (typeof $ === 'undefined') {
        console.warn('🎠 jQuery no está disponible, usando django.jQuery');
        $ = django.jQuery;
    }
    
    $(document).ready(function() {
        console.log('🎠 Carousel Admin JS loaded');
        
        // Mejorar el selector de color
        $('input[name="background_color"], input[name="text_color"]').each(function() {
            var $input = $(this);
            var $wrapper = $('<div class="color-picker-wrapper"></div>');
            $input.wrap($wrapper);
            
            // Crear input de tipo color
            var $colorPicker = $('<input type="color" class="color-picker">');
            $colorPicker.val($input.val());
            $input.after($colorPicker);
            
            // Sincronizar valores
            $colorPicker.on('change', function() {
                $input.val($(this).val());
                updateColorPreview($input, $(this).val());
            });
            
            $input.on('change', function() {
                $colorPicker.val($(this).val());
                updateColorPreview($input, $(this).val());
            });
            
            // Vista previa inicial
            updateColorPreview($input, $input.val());
        });
        
        // Función para actualizar vista previa de color
        function updateColorPreview($input, color) {
            var $preview = $input.siblings('.color-preview');
            if ($preview.length === 0) {
                $preview = $('<div class="color-preview"></div>');
                $input.after($preview);
            }
            
            $preview.css({
                'width': '30px',
                'height': '30px',
                'background-color': color,
                'border': '2px solid #ddd',
                'border-radius': '50%',
                'display': 'inline-block',
                'margin-left': '10px',
                'vertical-align': 'middle'
            });
        }
        
        // Validación de imagen
        $('input[name="image"]').on('change', function() {
            var file = this.files[0];
            if (file) {
                // Validar tipo de archivo
                var validTypes = ['image/jpeg', 'image/jpg', 'image/png', 'image/gif', 'image/webp'];
                if (!validTypes.includes(file.type)) {
                    alert('⚠️ Por favor selecciona un archivo de imagen válido (JPG, PNG, GIF, WebP)');
                    $(this).val('');
                    return;
                }
                
                // Validar tamaño (máximo 5MB)
                var maxSize = 5 * 1024 * 1024; // 5MB
                if (file.size > maxSize) {
                    alert('⚠️ La imagen es demasiado grande. Por favor selecciona una imagen menor a 5MB.');
                    $(this).val('');
                    return;
                }
                
                // Mostrar vista previa
                var reader = new FileReader();
                reader.onload = function(e) {
                    showImagePreview(e.target.result);
                };
                reader.readAsDataURL(file);
                
                console.log('✅ Imagen seleccionada:', file.name, 'Tamaño:', (file.size / 1024 / 1024).toFixed(2) + 'MB');
            }
        });
        
        // Función para mostrar vista previa de imagen
        function showImagePreview(src) {
            var $preview = $('#image-upload-preview');
            if ($preview.length === 0) {
                $preview = $('<div id="image-upload-preview"></div>');
                $('input[name="image"]').after($preview);
            }
            
            $preview.html(`
                <div style="margin-top: 15px; padding: 10px; background: #f8f9fa; border-radius: 5px; border: 1px solid #dee2e6;">
                    <p style="margin: 0 0 10px 0; font-weight: bold; color: #007bff;">📷 Vista previa de la imagen:</p>
                    <img src="${src}" style="max-width: 300px; max-height: 200px; object-fit: contain; border: 1px solid #ddd; border-radius: 5px; box-shadow: 0 2px 4px rgba(0,0,0,0.1);" />
                </div>
            `);
        }
        
        // Mejorar el mensaje de ayuda para imágenes
        var $imageField = $('.field-image');
        if ($imageField.length) {
            var helpText = `
                <div class="carousel-help-text">
                    <strong>💡 Consejos para subir imágenes:</strong><br>
                    • Tamaño recomendado: 1200x600 píxeles (ratio 2:1)<br>
                    • Formatos soportados: JPG, PNG, GIF, WebP<br>
                    • Tamaño máximo: 5MB<br>
                    • Para mejores resultados, usa imágenes de alta calidad
                </div>
            `;
            $imageField.append(helpText);
        }
        
        // Mejorar la interfaz de orden
        $('input[name="order"]').attr('min', '0').attr('step', '1').attr('placeholder', 'Ej: 1, 2, 3...');
        
        // Agregar tooltips informativos
        var tooltips = {
            'title': 'Título principal que aparecerá en el slide',
            'subtitle': 'Subtítulo opcional para información adicional',
            'description': 'Descripción detallada (opcional)',
            'button_text': 'Texto que aparecerá en el botón de acción',
            'button_link': 'URL a la que dirigirá el botón',
            'order': 'Orden de aparición (menor número = aparece primero)'
        };
        
        $.each(tooltips, function(fieldName, tooltip) {
            $(`input[name="${fieldName}"], textarea[name="${fieldName}"]`)
                .attr('title', tooltip)
                .attr('placeholder', tooltip);
        });
        
        console.log('✅ Carousel Admin JS initialized successfully');
    });
    
})(django.jQuery);